### PR TITLE
Assignment5+6 

### PR DESCRIPTION
**Pull request by mpennock5**
@Aedyn01

More test methods will be helpful.

### CODE REVIEW

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,30 +1,76 @@
...
 namespace Assignment
 {
     public class SampleData : ISampleData
     {
-        // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        private List<string>? _CsvRows;
+        private string peopleCSV = System.Reflection.Assembly.GetExecutingAssembly().Location.Replace("Assignment.dll", "People.csv");
+
+        public SampleData()
```
Comment (dcox2005) [inline Assignment/Assignment/SampleData.cs:13]: I passed the path into the constructor on the chance that the file is not located within the same location as the class file.

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,76 @@
...
+            split.Sort();
+
+            return split;
+        }
+           
 
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+            => string.Join(", ", GetUniqueSortedListOfStatesGivenCsvRows());
```
Comment (dcox2005) [inline Assignment/Assignment/SampleData.cs:45]: part 3 says to create a comma separated list, this is usually done with only and "," not ", ".  I would suggest doing this without the space.

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,76 @@
...
 
         // 4.
-        public IEnumerable<IPerson> People => throw new NotImplementedException();
+        public IEnumerable<IPerson> People
+        {
+            get
```
Comment (dcox2005) [inline Assignment/Assignment/SampleData.cs:50]: This is a property, I suggest creating it within the constructor.  The benefit to the way it is done here is that if the property is never called, the work for it is never done.  But if the property is called at least twice then the work is being done exponentially more.  We created a helper method that was called in the constructor and passed the results into this property.

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,76 @@
...
+                                              personAttribute[3]);
+                
+                return people;
+            }
+        }
 
         // 5.
         public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(
-            Predicate<string> filter) => throw new NotImplementedException();
+            Predicate<string> filter) =>
```
Comment (dcox2005) [inline Assignment/Assignment/SampleData.cs:66]: This looks like a more condensed and cleaner way to do this than the way we did it. Very nice

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Assignment.Tests;
```
Comment (dcox2005) [inline Assignment/Assignment.Tests/SampleDataTests.cs:8]: Tests names should be more meaning full. Ex "Part6_MethodReturnsUniqueList_SuccessListMatches". WhatIsBeingTested_HowItIsBeingTested_HowSuccessIsMeasured

**Review by dcox2005 — COMMENTED** [review]
Overall, very well done.  Some changes here and there that I would suggest, and some changes for my code I will probably implement :)

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of Spokane-based addresses. ❌  
            I did not see a test that tested this.
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌
          ToArray was never used.
Given the array, consider using string.Join to combine the list into a single string. ✔

Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔

Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
Use ISampleData.People for your data source. ✔

Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma separated list of states. ✔
Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use System.Linq.Enumerable.Aggregate LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

Given the implementation of Node in Assignment5
Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

File: Assignment/Assignment/SampleData.cs (hunk 3 above)
Comment (mpennock5) [reply Assignment/Assignment/SampleData.cs:50]: This is an excellent point, will definitely rework this and consider the implications of how its implemented. Good catch and thank you!

File: Assignment/Assignment/SampleData.cs [hunk 6]
```diff
@@ -1,30 +1,76 @@
...
     {
-        // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        private List<string>? _CsvRows;
+        private string peopleCSV = System.Reflection.Assembly.GetExecutingAssembly().Location.Replace("Assignment.dll", "People.csv");
+
+        public SampleData()
+        {
+            IEnumerable<string> CSV = File.ReadAllLines(peopleCSV);
+            CsvRows = CSV;
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:16]:
```suggestion
            _CsvRows = CSV;

```

File: Assignment/Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,76 @@
...
+
+        }
 
         // 2.
-        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            List<string> split = CsvRows.Select(x => x.Split(',')[6]).Distinct().ToList();
+            split.Sort();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:37]: You should use LINQs `OrderBy` for this. Additionally, you don't need to convert to a list with `ToList`.

File: Assignment/Assignment/SampleData.cs [hunk 8]
```diff
@@ -1,30 +1,76 @@
...
+            { 
+                IEnumerable<Person> people = from person in CsvRows.OrderBy(state => state.Split(',')[6])
+                .ThenBy(city => city.Split(',')[5]).ThenBy(zip => zip.Split(',')[7]).ToList()
+                let personAttribute = person.Split(',')
+                                              select new Person(personAttribute[1], 
+                                              personAttribute[2], 
+                                              new Address(personAttribute[4], personAttribute[5], personAttribute[6], personAttribute[7]),
+                                              personAttribute[3]);
+                
+                return people;
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:60]: When using LINQ, try to only use one of it's flavors (query or method flavor syntax)
The below is the solution using the method syntax (I reccommend this one as it is more C# like/looking)
```suggestion
                return CsvRows
                    .Select(x => x.Split(","))
                    .OrderBy(state => state[6])
                    .ThenBy(city => city[5])
                    .ThenBy(zip => zip[7])
                    .Select(personAttribute => new Person(
                        personAttribute[1],
                        personAttribute[2],
                        new Address(
                            personAttribute[4],
                            personAttribute[5],
                            personAttribute[6],
                            personAttribute[7]),
                        personAttribute[3])
                    );
```

**Pull request closed without merging**